src/mlpack/core/data/string_encoding_impl.hpp

```diff
@@ -0,0 +1,175 @@
+#define MLPACK_CORE_DATA_STRING_ENCODING_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "string_encoding.hpp"
+
+namespace mlpack {
+namespace data {
+
+void StringEncoding::Reset()
+{
+  mappings.clear();
+  originalStrings.clear();
+}
+
+StringEncoding::StringEncoding(const
+    StringEncoding& oldObject) : originalStrings(oldObject.originalStrings)
+{
+  std::deque<std::string>::iterator jt = originalStrings.begin();
+  for (auto it = oldObject.originalStrings.begin();
+      it != oldObject.originalStrings.end(); it++)
+  {
+    mappings[*jt] = oldObject.mappings.at(*it);
+    jt++;
+  }
+}
+
+StringEncoding& StringEncoding::operator= (
+    const StringEncoding &oldObject)
+{
+  if (this != &oldObject)
+  {
+    mappings.clear();
+    originalStrings.clear();
+    originalStrings = oldObject.originalStrings;
+    std::deque<std::string>::iterator jt = originalStrings.begin();
+    for (auto it = oldObject.originalStrings.begin();
+        it != oldObject.originalStrings.end(); it++)
+    {
+      mappings[*jt] = oldObject.mappings.at(*it);
+      jt++;
+    }
+  }
+  return *this;
+}
+
+template<typename TokenizerType>
+void StringEncoding::CreateMap(std::string& input,
+    TokenizerType tokenizer)
+{
+  boost::string_view strView(input);
+  boost::string_view token;
+  token = tokenizer(strView);
+  std::size_t curLabel = mappings.size() + 1;
+  while (!token.empty())
+  {
+    if (mappings.find(token) == mappings.end())
+    {
+        originalStrings.push_back(std::string(token));
+        mappings[originalStrings.back()] = curLabel++;
+    }
+    token = tokenizer(strView);
+  }
+}
+
+template<typename MatType, typename TokenizerType, typename EncodingPolicy>
+void StringEncoding::Encode(
+    const std::vector<std::string>& input,
+    MatType& output, TokenizerType tokenizer,
+    EncodingPolicy policy)
+{
+  boost::string_view strView;
+  boost::string_view token;
+  std::vector< std::vector<boost::string_view> > dataset;
+  size_t colSize = 0;
+  for (size_t i = 0; i < input.size(); i++)
+  {
+    strView = input[i];
+    token = tokenizer(strView);
+    dataset.push_back(std::vector<boost::string_view>() );
+    while (!token.empty())
+    {
+      dataset[i].push_back(token);
+      token = tokenizer(strView);
+    }
+    colSize = std::max(colSize, dataset[i].size());
+  }
+  size_t curLabel = mappings.size() + 1;
+  for (size_t i = 0; i < dataset.size(); ++i)
+  {
+    for (size_t j = 0; j < dataset[i].size(); ++j)
+    {
+      if (mappings.find(dataset[i][j]) == mappings.end())
+      {
+        originalStrings.push_back(std::string(dataset[i][j]));
+        mappings[originalStrings.back()] = curLabel++;
+      }
+    }
+  }
+  EncodingPolicy::InitMatrix(output, dataset.size(), colSize, mappings.size());
+  for (size_t i = 0; i < dataset.size(); ++i)
+    for (size_t j = 0; j < dataset[i].size(); ++j)
+        EncodingPolicy::Encode(mappings.at(dataset[i][j]), output, i, j);
```

Comment:
I tried to think about your question regarding  the use of vector<vector<size_t>> as the output for other policies. I suggest the solution:
We can add overloads of `InitMatrix()` and `Encode()` to each policy except `DictionaryEncodingPolicy`. These overloads should accept `vector<vector<size_t>>`. In this case the compiler won't throw any errors.

Reply:
Correct , Yeah that is a good solution, Wow :)